### PR TITLE
Remove Sock::Get() and Sock::Sock()

### PR DESCRIPTION
**Pull request by vasild**
_This is a piece of #21878, chopped off to ease review._

Peeking at the underlying socket file descriptor of `Sock` and checkig if it is `INVALID_SOCKET` is bad encapsulation and stands in the way of testing/mocking/fuzzing.

Instead use an empty `unique_ptr` to denote that there is no valid socket where appropriate or outright remove such checks where they are not necessary.

The default constructor `Sock::Sock()` is unnecessary now after recent changes, thus remove it.

### CODE REVIEW

**Comment by yancyribbens** [comment]
Why not add additional testing/mocking/fuzzing in this PR since this seems to be laying the groundwork for that?

**Comment by vasild** [comment]
@yancyribbens, to ease review. Right now, https://github.com/bitcoin/bitcoin/pull/21878 consists of this PR + the tests (it is almost complete :balloon:). If somebody has spare review cycles they can review/ACK #21878 directly.

**Comment by yancyribbens** [comment]
@vasild If the tests are not yet complete, then that may change this PR, right?  Besides adding more safety, seeing the accompanying tests would make this easier (for me anyway) to PR.

**Comment by vasild** [comment]
> If the tests are not yet complete, then that may change this PR, right?

The tests are "complete" in a sense that they are "production ready" and can be merged. "Not complete" in a sense that more tests can be added. I do not think that tweaking the tests can change this PR. The point of #21878 is to have high level code not deal with `int` sockets (file descriptor), but to deal with `Sock` instead, which can be mocked together with all the relevant operations (connect, accept, read, write, etc).

> Besides adding more safety, seeing the accompanying tests would make this easier (for me anyway) to review.

The tests are in #21878, feel free to look at them :heart:

**Comment by yancyribbens** [comment]
>  .. to have high level code not deal with int sockets (file descriptor), but to deal with Sock instead

Thanks for the additional explanation.  I was wondering if you could explain this a bit more.  It's been a while since i've done anything with int sockets but if I remember, it was in `C` (not C++).  Is `Sock` a C++ construct that makes it easier to Mock?

> The tests are in https://github.com/bitcoin/bitcoin/pull/21878, feel free to look at them heart

I looked over 21878 and I see there's some additional fuzz testing however I don't see the `Mocks` that this PR enables.  Am I missing the place where the Mocks are tested or is it yet to be added?  Would this PR be merged before 21878?

**Comment by yancyribbens** [comment]
@vasild 
> The problem with this is that in order to test it we have to create a real socket that really connects to the given address which is very inconvenient. It cannot be mocked.

I understand.  What I am asking is if you wouldn't mind adding the `Mock` tests to this PR, or create a new PR with the `Mocks` before merging this one?

**Comment by vasild** [comment]
> What I am asking is if you wouldn't mind adding the Mock tests to this PR,

I assume that by "the Mock tests" you mean the tests that are in the last 4 commits of https://github.com/bitcoin/bitcoin/pull/21878, right? If I add them to this PR then this PR would be the same as https://github.com/bitcoin/bitcoin/pull/21878.

I created this PR as a smaller piece of https://github.com/bitcoin/bitcoin/pull/21878 to ease review. If you want to peek or fully review the tests, then see the last 4 commits in https://github.com/bitcoin/bitcoin/pull/21878.

> or create a new PR with the Mocks before merging this one?

Because those tests depend on removing the `Sock::Get()` call which is done in this PR. If the tests get merged first, this may happen: `CConnman::OpenNetworkConnection()` -> `CConnman::ConnectNode()` -> `i2p::sam::Session::Connect()` and later `i2p::sam::Session::~Session()` -> `i2p::sam::Session::Disconnect()` -> `Sock::Get()`. Calling `Get()` on `FuzzedSock` is a bit fishy, the high level code shouldn't peek at the internal `int` socket and make decisions based on its value.

**Comment by yancyribbens** [comment]
@vasild 

Are there any Mock tests that are not part of fuzz tests?

**Comment by vasild** [comment]
There is one in `src/test/i2p_tests.cpp`.

**Comment by vasild** [comment]
It is already merged.

**Comment by yancyribbens** [comment]
I'm confused.  I thought this PR was required before a `Mock` could be used.

**Comment by vasild** [comment]
`FuzzedSock` and `StaticContentsSock` are mock implementations of `Sock` that are already used in `master`.

Depending on the code being tested various methods of `Sock` (or a mocked implementation) are going to be called. For example, if you write an unit or fuzz test for `CConnman::SocketSendData()` that is only going to invoke `Sock::Send()`.

**Comment by vasild** [comment]
`7b6fe6761a...666c7584cf`: rebase due to conflicts

**Comment by jonatack** [comment]
ACK 666c7584cfb5abe882909155ed0af76ead74ef3c modulo trivial rebase

**Comment by vasild** [comment]
`666c7584cf...c40c25763c`: rebase due to conflicts

Invalidates ACK from @jonatack

**Review by jonatack — COMMENTED** [review]
re-ACK c40c25763c6acc12e54824202b7bf503d3b2ae68 per `git range-diff dc905f6 666c758 c40c257`

It's good to have them in place, just noting that the `Sock::IsConnected()` virtual member function overrides in `src/test/util/net.h` and `src/test/fuzz/util/net.{h.cpp}` aren't used currently or here or in the parent yet.

**Comment by vasild** [comment]
`Sock::IsConnected()` is used in `src/i2p.cpp`. It is important to override all methods from the mocked sock implementations because when we exercise some real code from the tests with a mocked socket, then we do not want the real `IsConnected()` method to be called. For example, the tests from the [parent PR](https://github.com/bitcoin/bitcoin/pull/21878) mock the socket with `FuzzedSock` and call:

```
FUZZ_TARGET_INIT(connman, ...) which calls
  Connman::OpenNetworkConnection() which calls
    Connman::ConnectNode() which calls
      i2p::sam::Session::Connect() which calls
        i2p::sam::Session::CreateIfNotCreatedAlready() which calls
          Sock::IsConnected()  // must end up calling FuzzedSock::IsConnected()
```

**Comment by vasild** [comment]
`c40c25763c...e536b3a67b`: rebase due to conflicts

Invalidates ACK from @jonatack

**Review by theStack — COMMENTED** [review]
Concept ACK

File: src/test/sock_tests.cpp [hunk 1]
```diff
@@ -38,7 +38,6 @@ BOOST_AUTO_TEST_CASE(constructor_and_destructor)
 {
     const SOCKET s = CreateSocket();
     Sock* sock = new Sock(s);
-    BOOST_CHECK_EQUAL(sock->Get(), s);
```
Comment (ajtowns) [inline src/test/sock_tests.cpp:41]: Checking that you actually got the socket you wanted seems valuable (particularly for move operations). Maybe:

```c++
class Sock {
public:
    /** For unit tests */
    bool CheckSocket(SOCKET s) const { return s == m_socket; }
...
};

//  BOOST_CHECK_EQUAL(sock->Get(), s);
    BOOST_CHECK(sock->CheckSocket(s));
```

**Review by ajtowns — COMMENTED** [review]
Approach ACK

**Comment by vasild** [comment]
`e536b3a67b...b44e493890`: address suggestions

File: src/util/sock.h [hunk 2]
```diff
@@ -62,12 +61,6 @@ class Sock
      */
     virtual Sock& operator=(Sock&& other);
 
-    /**
-     * Get the value of the contained socket.
-     * @return socket or INVALID_SOCKET if empty
-     */
-    [[nodiscard]] virtual SOCKET Get() const;
```
[the remark this answers is not in the document]
Comment (vasild) [reply src/util/sock.h:69]: Done

File: src/util/sock.h [hunk 3]
```diff
@@ -26,11 +26,6 @@ static constexpr auto MAX_WAIT_FOR_IO = 1s;
 class Sock
 {
 public:
-    /**
-     * Default constructor, creates an empty object that does nothing when destroyed.
-     */
-    Sock();
```
[the remark this answers is not in the document]
Comment (vasild) [reply src/util/sock.h:33]: Done

File: src/test/sock_tests.cpp (hunk 1 above)
Comment (vasild) [reply src/test/sock_tests.cpp:41]: Done, but named it `operator==` instead of `CheckSocket`.

Comment (ajtowns) [reply src/test/sock_tests.cpp:41]: Could consider updating the commit message for `net: remove now unnecessary Sock::Get()`

Comment (vasild) [reply src/test/sock_tests.cpp:41]: Done.

**Comment by vasild** [comment]
`b44e493890...384e9e3552`: reword a commit message, see https://github.com/bitcoin/bitcoin/pull/26312#discussion_r1305495098

**Review by jonatack — COMMENTED** [review]
ACK 384e9e3552d7ee80b34bf695f1862884d6e34fd2

**Comment by vasild** [comment]
`384e9e3552...7df4508369`: add one more check in the unit test: https://github.com/bitcoin/bitcoin/pull/26312#discussion_r1306691259

File: src/test/sock_tests.cpp [hunk 4]
```diff
@@ -51,22 +51,33 @@ BOOST_AUTO_TEST_CASE(move_constructor)
...
     *sock2 = std::move(*sock1);
+    BOOST_CHECK(!SocketIsClosed(s1));
+    BOOST_CHECK(SocketIsClosed(s2));
+
     delete sock1;
-    BOOST_CHECK(!SocketIsClosed(s));
-    BOOST_CHECK_EQUAL(sock2->Get(), s);
+    BOOST_CHECK(!SocketIsClosed(s1));
+    BOOST_CHECK(SocketIsClosed(s2));
+    BOOST_CHECK(*sock2 == s1);
```
[the remark this answers is not in the document]
Comment (vasild) [reply src/test/sock_tests.cpp:77]: Added the second one.

**Comment by ajtowns** [comment]
ACK 7df450836969b81e98322c9a09c08b35d1095a25

**Comment by vasild** [comment]
Thank you! This concludes https://github.com/bitcoin/bitcoin/pull/21878 :white_check_mark:

Some more fuzz tests for review in https://github.com/bitcoin/bitcoin/pull/28584.

**Pull request merged**